[shard] Extensible ShardingSpec

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73309
* __->__ #72130

1. Refactor ShardingSpec, decouple PlacementSpec and ShardingSpec, as they are essentially two separate concept
2. Introduce customizable ShardingSpec, with the help of two APIs, we can allow user to inherit and define their own customized sharding spec:
  * ShardingSpec.build_metadata, which takes a tensor shape and define how to shard a tensor like this shape across ranks, return a ShardedTensorMetadata that describes the layout.
  * ShardingSpec.shard: define how to shard a tensor into ShardedTensor
3. Refactor `ShardedTensor.__init__` and `shard_parameter` to take the new ShardingSpec, enable these two API to support both ChunkShardingSpec and EnumerableShardingSpec

Differential Revision: [D33923403](https://our.internmc.facebook.com/intern/diff/D33923403/)

Commit:
[shard] Extensible ShardingSpec

...1. Refactor ShardingSpec, decouple PlacementSpec and ShardingSpec, as they are essentially two separate concept
2. Introduce ShardingSpec.shard, which takes a tensor shape and define how to shard a tensor like this shape across ranks, return a list of ShardMetadata that describes the layout.
3. Refactor `ShardedTensor.__init__` and `shard_parameter` to take the new ShardingSpec, enable these two API to support both ChunkShardingSpec and EnumerableShardingSpec

TODO:
* make utils like `_parse_and_validate_remote_device` accesible within the sharding_spec package
* add tests to test a custom sharding spec.

Differential Revision: [D33923403](https://our.internmc.facebook.com/intern/diff/D33923403/)

[ghstack-poisoned]

fduwjj commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -110,3 +165,8 @@ def __post_init__(self):
 
         validate_non_overlapping_shards_metadata(self.shards)
+
+    def shard(self, tensor_sizes: List[int], process_group=None) -> List[ShardMetadata]:
+        # check if shards form a valid tensor
+        check_tensor(self.shards, tensor_sizes)
+        return self.shards
```
N00b question, are we going to do the sharding here? Or this is actually just a no-op plus validation check?

wanchaol replied:
actually after some discussion, I think we might not able to do sharding for EnumerableShardingSpec as we don't have the clue how the shards layout look like and how to scatter the shards, just left this not implemented until we figure out a better way

fduwjj reviewed (commented):
Thanks for the prototyping, it looks good to me overall. Do you mind rebasing on top of https://github.com/pytorch/pytorch/pull/72087 and https://github.com/pytorch/pytorch/pull/70079?

Thanks!

Commit:
Update on "[shard] Extensible ShardingSpec"


...1. Refactor ShardingSpec, decouple PlacementSpec and ShardingSpec, as they are essentially two separate concept
2. Introduce ShardingSpec.shard, which takes a tensor shape and define how to shard a tensor like this shape across ranks, return a list of ShardMetadata that describes the layout.
3. Refactor `ShardedTensor.__init__` and `shard_parameter` to take the new ShardingSpec, enable these two API to support both ChunkShardingSpec and EnumerableShardingSpec

TODO:
* make utils like `_parse_and_validate_remote_device` accesible within the sharding_spec package
* add tests to test a custom sharding spec.

Differential Revision: [D33923403](https://our.internmc.facebook.com/intern/diff/D33923403/)

[ghstack-poisoned]

Commit:
Update on "[shard] Extensible ShardingSpec"


...1. Refactor ShardingSpec, decouple PlacementSpec and ShardingSpec, as they are essentially two separate concept
2. Introduce ShardingSpec.shard, which takes a tensor shape and define how to shard a tensor like this shape across ranks, return a list of ShardMetadata that describes the layout.
3. Refactor `ShardedTensor.__init__` and `shard_parameter` to take the new ShardingSpec, enable these two API to support both ChunkShardingSpec and EnumerableShardingSpec

TODO:
* make utils like `_parse_and_validate_remote_device` accesible within the sharding_spec package
* add tests to test a custom sharding spec.

Differential Revision: [D33923403](https://our.internmc.facebook.com/intern/diff/D33923403/)

[ghstack-poisoned]

kumpera commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -84,6 +105,40 @@ def _verify_dim(dim):
         if not (isinstance(dim, int) or isinstance(dim, str)):
             raise ValueError(f'{dim} needs to either be an int or str')
 
+    def shard(self, tensor_sizes: List[int], process_group=None) -> List[ShardMetadata]:
```
This API design forces us to have a process group and have process groups enabled?
The issue is that those constraints make it not usable for planning or testing purposes.

This API should allow us to answer questions like "how different is rank 0 usage if I go from 14 to 12 workers?"

wanchaol replied:
I see, yeah good point. I updated the PR with two new APIs now:
1. ShardingSpec.build_metadata, this will return a ShardedTensorMetadata, I will remove the process_group argument as it seems not very useful. This can be used by planning or testing purposes. 
2. ShardingSpec.shard, this will return an actual ShardedTensor, which will have the process_group as argument as we really scattering the shards to different ranks.

For answering the question like "how different is rank 0 usage if I go from 14 to 12 workers?" I feel like this might be the responsibility of a "ShardingPlanner", ShardingSpec should just try to build the metadata base on the spec (list of placements) given. ShardingPlanner can explore different options to make use of the resources (i.e. device usage). We will build a Customizable ShardingPlanner as described in this RFC https://github.com/pytorch/pytorch/issues/72138, as user can define their own rules to explore their cluster set up.

Commit:
Update on "[shard] Extensible ShardingSpec"


...1. Refactor ShardingSpec, decouple PlacementSpec and ShardingSpec, as they are essentially two separate concept
2. Introduce ShardingSpec.shard, which takes a tensor shape and define how to shard a tensor like this shape across ranks, return a list of ShardMetadata that describes the layout.
3. Refactor `ShardedTensor.__init__` and `shard_parameter` to take the new ShardingSpec, enable these two API to support both ChunkShardingSpec and EnumerableShardingSpec

TODO:
* make utils like `_parse_and_validate_remote_device` accesible within the sharding_spec package
* add tests to test a custom sharding spec.

Differential Revision: [D33923403](https://our.internmc.facebook.com/intern/diff/D33923403/)

[ghstack-poisoned]

Commit:
Update on "[shard] Extensible ShardingSpec"


...1. Refactor ShardingSpec, decouple PlacementSpec and ShardingSpec, as they are essentially two separate concept
2. Introduce ShardingSpec.shard, which takes a tensor shape and define how to shard a tensor like this shape across ranks, return a list of ShardMetadata that describes the layout.
3. Refactor `ShardedTensor.__init__` and `shard_parameter` to take the new ShardingSpec, enable these two API to support both ChunkShardingSpec and EnumerableShardingSpec

TODO:
* make utils like `_parse_and_validate_remote_device` accesible within the sharding_spec package
* add tests to test a custom sharding spec.

Differential Revision: [D33923403](https://our.internmc.facebook.com/intern/diff/D33923403/)

[ghstack-poisoned]

Commit:
Update on "[shard] Extensible ShardingSpec"


...1. Refactor ShardingSpec, decouple PlacementSpec and ShardingSpec, as they are essentially two separate concept
2. Introduce ShardingSpec.shard, which takes a tensor shape and define how to shard a tensor like this shape across ranks, return a list of ShardMetadata that describes the layout.
3. Refactor `ShardedTensor.__init__` and `shard_parameter` to take the new ShardingSpec, enable these two API to support both ChunkShardingSpec and EnumerableShardingSpec

TODO:
* make utils like `_parse_and_validate_remote_device` accesible within the sharding_spec package
* add tests to test a custom sharding spec.

Differential Revision: [D33923403](https://our.internmc.facebook.com/intern/diff/D33923403/)

[ghstack-poisoned]

fduwjj reviewed (commented):
This looks great and thanks for helping us get rid of writing duplicate metadata code. Just have some questions about reshard.

Commit:
Update on "[shard] Extensible ShardingSpec"


...1. Refactor ShardingSpec, decouple PlacementSpec and ShardingSpec, as they are essentially two separate concept
2. Introduce ShardingSpec.shard, which takes a tensor shape and define how to shard a tensor like this shape across ranks, return a list of ShardMetadata that describes the layout.
3. Refactor `ShardedTensor.__init__` and `shard_parameter` to take the new ShardingSpec, enable these two API to support both ChunkShardingSpec and EnumerableShardingSpec

TODO:
* make utils like `_parse_and_validate_remote_device` accesible within the sharding_spec package
* add tests to test a custom sharding spec.

Differential Revision: [D33923403](https://our.internmc.facebook.com/intern/diff/D33923403/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -611,11 +545,11 @@ def reshard(self, resharding_spec: ShardingSpec) -> ShardedTensor:
+        #     not isinstance(resharding_spec, ChunkShardingSpec) or
+        #     not isinstance(self._sharding_spec, ChunkShardingSpec)
+        # ):
+        #     raise NotImplementedError("Only ChunkShardingSpec supported for reshard.")
```
Yeah you are right, I just reverted this, I guess we can make this generic, but i'm not sure how hard this would be in terms of resharding with Enumerable or other sharding spec.

wanchaol commented on test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:
```diff
@@ -1787,32 +1784,22 @@ def test_with_rpc_names(self):
+                local_tensor = torch.rand(shard_metadata.shard_sizes).cuda(device)
+                local_shard_metadata = shard_metadata
+
+        # TODO: figure out what the API should behave when some rank have no shard
```
made some changes in test about `init_from_local_tensor`, to reflect our offline discussion

fduwjj replied:
sounds good to me.

fduwjj reviewed (approved):
lgtm and please make sure linter is all cleared.

Commit:
Update on "[shard] Extensible ShardingSpec"


...1. Refactor ShardingSpec, decouple PlacementSpec and ShardingSpec, as they are essentially two separate concept
2. Introduce ShardingSpec.shard, which takes a tensor shape and define how to shard a tensor like this shape across ranks, return a list of ShardMetadata that describes the layout.
3. Refactor `ShardedTensor.__init__` and `shard_parameter` to take the new ShardingSpec, enable these two API to support both ChunkShardingSpec and EnumerableShardingSpec

TODO:
* make utils like `_parse_and_validate_remote_device` accesible within the sharding_spec package
* add tests to test a custom sharding spec.

Differential Revision: [D33923403](https://our.internmc.facebook.com/intern/diff/D33923403/)

[ghstack-poisoned]

Commit:
Update on "[shard] Extensible ShardingSpec"


1. Refactor ShardingSpec, decouple PlacementSpec and ShardingSpec, as they are essentially two separate concept
2. Introduce customizable ShardingSpec, with the help of two APIs, we can allow user to inherit and define their own customized sharding spec:
  * ShardingSpec.build_metadata, which takes a tensor shape and define how to shard a tensor like this shape across ranks, return a ShardedTensorMetadata that describes the layout.
  * ShardingSpec.shard: define how to shard a tensor into ShardedTensor
3. Refactor `ShardedTensor.__init__` and `shard_parameter` to take the new ShardingSpec, enable these two API to support both ChunkShardingSpec and EnumerableShardingSpec

Differential Revision: [D33923403](https://our.internmc.facebook.com/intern/diff/D33923403/)

[ghstack-poisoned]

Commit:
Update on "[shard] Extensible ShardingSpec"


1. Refactor ShardingSpec, decouple PlacementSpec and ShardingSpec, as they are essentially two separate concept
2. Introduce customizable ShardingSpec, with the help of two APIs, we can allow user to inherit and define their own customized sharding spec:
  * ShardingSpec.build_metadata, which takes a tensor shape and define how to shard a tensor like this shape across ranks, return a ShardedTensorMetadata that describes the layout.
  * ShardingSpec.shard: define how to shard a tensor into ShardedTensor
3. Refactor `ShardedTensor.__init__` and `shard_parameter` to take the new ShardingSpec, enable these two API to support both ChunkShardingSpec and EnumerableShardingSpec

Differential Revision: [D33923403](https://our.internmc.facebook.com/intern/diff/D33923403/)

[ghstack-poisoned]

pritamdamania87 reviewed (commented):
Late review on this, just have a few minor comments. Feel free to land this PR and address them in follow up PRs.

wanchaol commented on test/distributed/_shard/sharding_spec/test_sharding_spec.py:
```diff
@@ -355,5 +375,134 @@ def test_infer_sharding_spec_from_shards_metadata(self):
+              src_rank: int = 0,
+              process_group=None) -> ShardedTensor:
+
+        raise NotImplementedError("GridShardingSpec.shard not implemented yet!")
```
Yeah sure I will do a follow up diff to test it, I just raise an error here bc i think it's sufficient to test that `shard_parameter` is going through this route. How to slice the tensor and scatter the "grids" could be a bit tricker to implement haha.

wanchaol commented on torch/distributed/_shard/sharded_tensor/__init__.py:
```diff
@@ -3,31 +3,24 @@
 
 
-def empty(sharding_spec: ShardingSpec,
+def empty(sharding_spec: shard_spec.ShardingSpec,
```
This is purely bc of recursive dependency, we use `ShardedTensor` in `sharding_spec` package, so we can't initialize `sharding_spec` here, otherwise it will trigger recursive imports.
